### PR TITLE
feat: enable table and column names to contain spaces. closes #419

### LINKED ISSUE #419: UMCG Data Catalogue wants to have spaces the name of items in forms
**Is your feature request related to a problem? Please describe.**
UMCG Data Catalogue people want to be able to change the names of the items in forms, e.g. ageMin should be Minimum age.

**Describe the solution you'd like**
Add a label to the data model.
=> we will not add labels, however, we will enable users to add spaces to their names so they can say 'Minimum Age'.

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by mswertz**
You can use spaces in names of tables and columns.

To enable this the 'graphql' api has been altered, and for each 'name' also an escaped 'id' is generated.
* In case a table or column name contains a space then it is replace with '_'
* In case a table contains multiple consequetive space it is also replaced with one '_'
* Every '_' is replace by '__'
* substring '_ ' or '_ ' is prohibited in names

As a consequence of this change, also the client javascript needed to use the 'id' instead of 'name' in context of graphql requests and json data.

closes #419

### CODE REVIEW

**Comment by sidohaakma** [comment]
[feat: enable table and column names to contain spaces. closes #419](https://trello.com/c/V0mdz6Mc/302-feat-enable-table-and-column-names-to-contain-spaces-closes-419)

File: backend/molgenis-emx2-graphql/src/test/java/org/molgenis/emx2/graphql/TestGraphqSchemaFields.java [hunk 1]
```diff
@@ -400,6 +401,34 @@ public void testAddAlterDropColumn() throws IOException {
...
+
+    int count = execute("{Persondetails_agg{count}}").at("/Persondetails_agg/count").intValue();
+
+    //    // insert should increase count
+    //    execute("mutation{insert(Tag:{name:\"blaat\"}){message}}");
+    //    TestCase.assertEquals(count + 1,
+    // execute("{Tag_agg{count}}").at("/Tag_agg/count").intValue());
+    //    // delete
+    //    execute("mutation{delete(Tag:{name:\"blaat\"}){message}}");
+    //    TestCase.assertEquals(count, execute("{Tag_agg{count}}").at("/Tag_agg/count").intValue());
```
Comment (connoratrug) [inline backend/molgenis-emx2-graphql/src/test/java/org/molgenis/emx2/graphql/TestGraphqSchemaFields.java:426]: remove commented code

**Comment by connoratrug** [comment]
Ssuggest tot add some description to the PR that summarises the problem and selected solution. Domain objects ( table , column) get id field added ( generated from name) this should probably be added to the docs as the changes the way the graphql should be used

File: backend/molgenis-emx2-graphql/src/main/java/org/molgenis/emx2/graphql/GraphqlApiFactory.java [hunk 2]
```diff
@@ -101,13 +104,44 @@ public static String convertExecutionResultToJson(ExecutionResult executionResul
   /** bit unfortunate that we have to convert from json to map and back */
   static Object transform(String json) throws IOException {
     // benchmark shows this only takes a few ms so not a large performance issue
+    // alternatively, we should change the SQL to result escaped results but that is a nightmare to
```
Comment (connoratrug) [inline backend/molgenis-emx2-graphql/src/main/java/org/molgenis/emx2/graphql/GraphqlApiFactory.java:107]: 'should' or 'could'

**Comment by connoratrug** [comment]
remove debug code from ui

Screenshot 2021-11-10 at 09 58 02

**Comment by connoratrug** [comment]
[Allow white space in table name and column name](https://trello.com/c/BGrgy1A1/308-allow-white-space-in-table-name-and-column-name)

File: backend/molgenis-emx2-graphql/src/main/java/org/molgenis/emx2/graphql/GraphqlApiFactory.java (hunk 2 above)
Comment (mswertz) [reply backend/molgenis-emx2-graphql/src/main/java/org/molgenis/emx2/graphql/GraphqlApiFactory.java:107]: what do you think? I think if it becomes slow we 'should'

**Comment by mswertz** [comment]
> remove debug code from ui

fixed

**Review by connoratrug — CHANGES_REQUESTED** [review]
Cannot add row anymore

- create db, create table and column , add row ; -> ```[ { "message": "Error: The variables input contains a field name 'mg_draft' that is not defined for input object type 'tableaInput' " } ]```

File: backend/molgenis-emx2/src/main/java/org/molgenis/emx2/Column.java [hunk 3]
```diff
@@ -67,13 +67,17 @@ public Column(String columnName) {
...
-    if (!skipValidation && !columnName.matches("[a-zA-Z][a-zA-Z0-9_]*")) {
+    if (!skipValidation && !columnName.matches("[a-zA-Z][a-zA-Z0-9_ ]*")) {
       throw new MolgenisException(
           "Invalid column name '"
               + columnName
-              + "': Column must start with a letter, followed by letters, underscores or numbers, i.e. [a-zA-Z][a-zA-Z0-9_]*");
+              + "': Column must start with a letter, followed by letters, underscores, a space or numbers, i.e. [a-zA-Z][a-zA-Z0-9_]*");
     }
-    this.columnName = columnName;
+    if (!skipValidation && (columnName.contains("_ ") || columnName.contains(" _"))) {
```
[the remark this answers is not in the document]
Comment (mswertz) [reply backend/molgenis-emx2/src/main/java/org/molgenis/emx2/Column.java:76]: actually, these cases are being catched. Will add a test to demonstrate

File: backend/molgenis-emx2-graphql/src/main/java/org/molgenis/emx2/graphql/GraphqlTableFieldFactory.java [hunk 4]
```diff
@@ -455,10 +470,22 @@ private Filter convertMapToFilter(String name, Map<String, Object> subFilter) {
           }
           result.add(sc);
         }
+      }
     }
     return result.toArray(new SelectColumn[result.size()]);
   }
 
+  private Optional<Column> findColumn(TableMetadata aTable, String name) {
```
[the remark this answers is not in the document]
Comment (mswertz) [reply backend/molgenis-emx2-graphql/src/main/java/org/molgenis/emx2/graphql/GraphqlTableFieldFactory.java:478]: good suggestion

File: backend/molgenis-emx2-graphql/src/main/java/org/molgenis/emx2/graphql/GraphqlTableFieldFactory.java [hunk 5]
```diff
@@ -90,26 +90,26 @@ public GraphQLFieldDefinition tableAggField(Table table) {
   }
 
   private GraphQLObjectType createTableObjectType(Table table) {
-    GraphQLObjectType.Builder tableBuilder = GraphQLObjectType.newObject().name(table.getName());
-    for (Column col : table.getMetadata().getColumnsWithoutHeadings())
+    GraphQLObjectType.Builder tableBuilder =
+        GraphQLObjectType.newObject().name(escape(table.getName()));
+    for (Column col : table.getMetadata().getColumnsWithoutHeadings()) {
+      String name = escape(col.getName());
```
[the remark this answers is not in the document]
Comment (mswertz) [reply backend/molgenis-emx2-graphql/src/main/java/org/molgenis/emx2/graphql/GraphqlTableFieldFactory.java:96]: ok

**Review by connoratrug — APPROVED** [review]

**Pull request merged**